draft-ietf-quic-transport.md

```diff
@@ -1448,6 +1448,32 @@ New transport parameters can be registered according to the rules in
 {{iana-transport-parameters}}.
 
 
+## Cryptographic Message Buffering
+
+Implementations need to maintain a buffer of CRYPTO data received out of order.
+Because there is no flow control of CRYPTO frames, an endpoint could
+potentially force its peer to buffer an unbounded amount of data.
+
+Implementations MUST support buffering at least 4096 bytes of data received out
+of order in CRYPTO frames. Endpoints MAY choose to allow more data to be
+buffered during the handshake. A larger limit during the handshake could allow
+for larger keys or credentials to be exchanged. An endpoint's buffer size does
+not need to remain constant during the life of a connection.
+
+Being unable to buffer CRYPTO frames during the handshake can lead to a
+connection failure. If its buffer is exceeded during the handshake, an endpoint
+can temporarily expand its buffer to complete the handshake. If an endpoint
+cannot expand its buffer, it MUST close the connection with a
+CRYPTO_BUFFER_EXCEEDED error code.
+
+Once the handshake completes, if an endpoint is unable to buffer all data in a
+CRYPTO frame, it MAY discard all subsequent CRYPTO frames, or it MAY close the
```

Comment:
"subsequent" here could mean further in the CRYPTO stream, or could mean received later in time.  If the latter, then the currently-buffered crypto data will never be delivered and can also be discarded; we should say so.  In either case, we should be specific about which "subsequent" we mean.

Reply:
I didn't consider it potentially meaning further in the CRYPTO stream (I was thinking later in time). I've clarified that it means later in time.

Previously I had added language to state that if CRYPTO frames are being discarded, then the endpoint can tear down all crypto state that it has. I removed that since @martinthomson pointed out that's an obvious thing to do. I'm thinking that discarding any buffered but not processed crypto data falls under tearing down crypto state. If this no longer seems obvious, I'll add back that the endpoint can tear down crypto state and discard any remaining buffered crypto data.